Integrations/MicrosoftGraphGroups/MicrosoftGraphGroups.py

```diff
@@ -0,0 +1,441 @@
+            parsed_response = dbot_response.json()
+        except ValueError:
+            raise Exception(
+                'There was a problem in retrieving an updated access token.\n'
+                'The response from the Demistobot server did not contain the expected content.'
+            )
+        access_token = parsed_response.get('access_token')
+        expires_in = parsed_response.get('expires_in', 3595)
+        time_now = epoch_seconds()
+        time_buffer = 5  # seconds by which to shorten the validity period
+        if expires_in - time_buffer > 0:
+            # err on the side of caution with a slightly shorter access token validity period
+            expires_in = expires_in - time_buffer
+
+        demisto.setIntegrationContext({
+            'access_token': access_token,
+            'valid_until': time_now + expires_in
+        })
+        return access_token
+
+    def http_request(self, method: str, url_suffix: str = None, params: Dict = None, body: Optional[str] = None,
+                     next_link: str = None):
+        """
+        Generic request to Microsoft Graph
+        """
+        token = self.get_access_token()
+        if next_link:
+            url = next_link
+        else:
+            url = f'{self.base_url}{url_suffix}'
+
+        response = requests.request(
+            method,
+            url,
+            headers={
+                'Authorization': 'Bearer ' + token,
+                'Content-Type': 'application/json',
+                'Accept': 'application/json'
+            },
+            params=params,
+            data=body,
+            verify=self.use_ssl,
+        )
+        try:
+            data = response.json() if response.text else {}
+            if not response.ok:
+                return_error(f'API call to MS Graph failed [{response.status_code}]'
+                             f' - {demisto.get(data, "error.message")}')
+            elif response.status_code == 206:  # 206 indicates Partial Content, reason will be in the warning header
+                demisto.debug(str(response.headers))
+
+            return data
+
+        except TypeError as ex:
+            demisto.debug(str(ex))
+            return_error(f'Error in API call to Microsoft Graph, could not parse result [{response.status_code}]')
+
+    def test_function(self):
+        token = self.get_access_token()
+        response = requests.get(
+            self.base_url + 'groups',
+            headers={
+                'Authorization': 'Bearer ' + token,
+                'Content-Type': 'application/json',
+                'Accept': 'application/json'
+            },
+            params={'$orderby': 'displayName'},
+            verify=self.use_ssl
+        )
+        try:
+            data = response.json() if response.text else {}
+            if not response.ok:
+                return_error(f'API call to MS Graph failed. Please check authentication related parameters.'
```

Comment:
you should raise formatted exception instead of return error. this should be handled in the main func.